src/test/scala/ch/epfl/pop/tests/json/JsonMessageParserTest.scala

```diff
@@ -27,6 +30,19 @@ class JsonMessageParserTest extends FunSuite with Matchers {
         }
       }
 
+      @scala.annotation.tailrec
+      def checkListOfKeySigPair(l1: List[KeySignPair], l2: List[KeySignPair]): Unit = {
+        l1.length should equal (l2.length)
+
+        (l1, l2) match {
+          case _ if l1.isEmpty =>
+          case (h1 :: tail1, h2 :: tail2) =>
+            h1.witness should equal (h2.witness)
+            h1.signature should equal (h2.signature)
+            checkListOfKeySigPair(tail1, tail2)
+        }
+      }
+
 
       val o1 = this.m
       val o2 = o
```

Comment:
What is o1 and o2?

Reply:
Yes! Very good question! The important parts around those two lines resemble this:

```scala
implicit class RichJsonMessage(m: JsonMessage) {
    def shouldBeEqualUntilMessageContent(o: JsonMessage): Unit = {

        def checkListOfByteArray(...) { /* ... */ }
        def checkListOfKeySigPair(...) { /* ... */ }

        val o1 = this.m
        val o2 = o

        o1 match {
            // actual tests
        }
    }
}
```

The whole point of the `RichJsonMessage` implicit class is to implicitly transform a `JsonMessage` into a `RichJsonMessage`. This way, I can use the method `shouldBeEqualUntilMessageContent` on a `JsonMessage` without adding this method in the real code (not test cases).

Thus, when later I use

```scala
m1 shouldBeEqualUntilMessageContent m2
// m1 and m2 are both standard JsonMessage
```

then `m1` gets implicitly extended into its Rich version (argument `m` of `RichJsonMessage`) and the method is called on it using `m2` as argument (thus `o` is `m2`)

To sum it all up, we have

```scala
val o1 = this.m    // corresponds to m1 due to implicit class conversion (parameter of the class)
val o2 = o         // corresponds to m2 (parameter of the method)
```

Reply:
You may have better names for them.

Reply:
Alrigth! I will merge this PR and do this in a little PR addressing bug fixes tonight or tomorrow. If I add more code, the action that we pushed will now be run on this code and it will fail